docs/advanced_topics.rst

```diff
@@ -0,0 +1,43 @@
+
+Thread affinity
+---------------
+
+The rationale behind this functionality it to allow pining DALI threads to given CPU. It can save the overhead of worker threads jumping from core to core and increase performance for CPU heave workloads.
+It is possible to set CPU thread affinity for DALI using ``DALI_AFFINITY_MASK`` environment variable. ``DALI_AFFINITY_MASK`` should be a comma separated list of CPU ids numbers that will be assigned to corresponding DALI threads. A number of DALI threads is set during pipeline construction by ``num_threads`` argument, ``set_affinity`` enables setting thread affinity for the CPU worker threads. It is worth to mention that due to performance reason nvJPEG decoder operator houses threads on its own and they are always affined.
+When ``DALI_AFFINITY_MASK`` is empty affinity is set based on the result returned by nvmlDeviceGetCpuAffinity.
+If there are more threads than values in ``DALI_AFFINITY_MASK`` value from nvmlDeviceGetCpuAffinity will be used for a thread without value provided by an environment variable,
+I.e.:
+
+.. code-block:: bash
+
+  num_threads=5
+  DALI_AFFINITY_MASK=3,5,6,10
+
+Will set thread 0 to CPU 3, 1 to CPU 5, 2 to CPU 6, 3 to CPU 10 and 4 to CPU which is returned by nvmlDeviceGetCpuAffinity.
+
+Memory consumption
+------------------
+
+DALI architecture makes all GPU operators processing the whole batch at the time and CPU one sample at the time. As DALI is only enlarging its memory, never shrinking, each GPU buffer is as large as the largest possible batch, while CPU buffers are as large as batch size multiplied by the size of the largest sample (even though CPU processes one sample at the time per thread still vector of samples need to reside in the memory). This lazy allocation strategy helps in cases when memory allocation is expensive like pinned CPU memory or GPU memory reducing the number of total memory operations. This is most visible for the operators which output size may differ from sample to sample, from run to run. Operator with the fixed size outputs - like crop is not affecting overall memory consumption growth over time.
+
+Operator buffer presizing
+-------------------------
+
+The rationale behind this functionality is allowing the user to fine tune his processing pipeline when he can precisely forecast memory consumption during DALI run saving the overhead of some reallocations.
+DALI uses intermediate buffers to pass data between operators in the processing graph. Due to internal DALI architecture, memory is newer freed but just enlarged when present buffers are not able to house data they need to fit in. However, in some cases, even this limited number of allocations still could affect DALI performance. If the user knows how much memory each operator buffer may need, it is possible to provide a hint to presize buffers before the first run.
```

Comment:
The purpose of this functionality is to enable the user to fine-tune the processing pipeline in situations when it is possible to forecast precisely the memory consumption during a DALI run. This results in saving the overhead of some reallocations.

DALI uses intermediate buffers to pass data between operators in the processing graph. With DALI, the memory is never freed but just enlarged when present buffers are not sufficient to hold the data. However, in some cases, even this limited number of allocations still could affect DALI performance. Hence, if the user knows how much memory each operator buffer needs, then it is possible to provide a hint to presize buffers before the first run.

Reply:
Adjusted